adal/src/telemetry/java/com/microsoft/aad/adal/HttpEvent.java

```diff
@@ -87,20 +90,108 @@ void setRequestIdHeader(final String requestIdHeader) {
         setProperty(EventStrings.REQUEST_ID_HEADER, requestIdHeader);
     }
 
+    void setSpeRingInfo(final String xMsCliTelem) {
+        // if the header isn't present, do nothing
+        if (StringExtensions.isNullOrBlank(xMsCliTelem)) {
+            return;
+        } else {
+            Logger.d(TAG, "Parsing x-ms-clitelem header: " + xMsCliTelem);
+        }
+
+        // split the header based on the delimiter
+        final String[] headerSegments = xMsCliTelem.split(",");
+
+        // make sure the header isn't empty
+        if (0 == headerSegments.length) {
+            Logger.w(TAG, "SPE Ring header missing version field.", null, ADALError.X_MS_CLITELEM_VERSION_UNRECOGNIZED);
+            return;
+        }
+
+        // get the version of this header
+        final String headerVersion = headerSegments[0];
+
+        // declare values tracked by this header
+        String errorCode = null;
+        String subErrorCode = null;
+        String tokenAge = null;
+        String speRing = null;
+
+        if (headerVersion.equals("1")) {
+            final int indexErrorCode = 1;
+            final int indexSubErrorCode = 2;
+            final int indexTokenAge = 3;
+            final int indexSpeInfo = 4;
+
+            // get the error_code
+            if (headerSegments.length >= indexErrorCode + 1) {
+                errorCode = headerSegments[indexErrorCode];
+            }
+
+            // get the sub_error_code
+            if (headerSegments.length >= indexSubErrorCode + 1) {
+                subErrorCode = headerSegments[indexSubErrorCode];
+            }
+
+            // get the token_age
+            if (headerSegments.length >= indexTokenAge + 1) {
+                tokenAge = headerSegments[indexTokenAge];
+            }
+
+            // get the spe_ring
+            if (headerSegments.length >= indexSpeInfo + 1) {
+                speRing = headerSegments[indexSpeInfo];
+            }
```

Comment:
in iOS implementation I was a little stricter about it. I assume that if version number is 1 as expected, then I check that there's exactly N number of elements present (5). If there's less or more elements, data will be considered invalid and discarded. With this implementation, if we get e.g "1,1," we still try to parse as many elements as possible, but according to spec there should be exactly N elements for this version in any case and some of them might be empty. I'm not sure, which approach we should decide to go with, but let's agree to use the same for both platforms.

Reply:
Sounds good. See 5e0ecd2a2056470bfb06886d0e4c634cb91abbc2

Reply:
Added new tests for this, too 👌